components/raftstore/src/store/async_io/write.rs

```diff
@@ -0,0 +1,839 @@
+                }
+            }
+
+            let now = Instant::now();
+            STORE_WRITE_HANDLE_MSG_DURATION_HISTOGRAM.observe(duration_to_sec(
+                now.saturating_duration_since(handle_begin)
+                    .saturating_sub(after_persist),
+            ));
+
+            if self.trans.need_flush() {
+                self.trans.flush();
+                self.message_metrics.flush();
+                STORE_WRITE_FLUSH_SEND_DURATION_HISTOGRAM
+                    .observe(duration_to_sec(now.saturating_elapsed()));
+            }
+
+            if self.batch.is_empty() {
+                continue;
+            }
+
+            STORE_WRITE_TRIGGER_SIZE_HISTOGRAM.observe(self.batch.get_raft_size() as f64);
+
+            self.write_to_db();
+
+            self.metrics.flush();
+
+            STORE_WRITE_LOOP_DURATION_HISTOGRAM
+                .observe(duration_to_sec(handle_begin.saturating_elapsed()));
+        }
+    }
+
+    fn write_to_db(&mut self) {
+        self.batch.before_write_to_db(&self.metrics);
+
+        fail_point!("raft_before_save");
+
+        if !self.batch.kv_wb.is_empty() {
+            let now = Instant::now();
+            let mut write_opts = WriteOptions::new();
+            write_opts.set_sync(true);
+            self.batch.kv_wb.write_opt(&write_opts).unwrap_or_else(|e| {
+                panic!(
+                    "store {}: {} failed to write to kv engine: {:?}",
+                    self.store_id, self.tag, e
+                );
+            });
+            if self.batch.kv_wb.data_size() > KV_WB_SHRINK_SIZE {
+                self.batch.kv_wb = self.engines.kv.write_batch_with_cap(KV_WB_DEFAULT_SIZE);
+            }
+            STORE_WRITE_KVDB_DURATION_HISTOGRAM
+                .observe(duration_to_sec(now.saturating_elapsed()) as f64);
+        }
+
+        self.batch.after_write_to_kv_db(&self.metrics);
+
+        fail_point!("raft_between_save");
+
+        if !self.batch.raft_wb.is_empty() {
+            fail_point!("raft_before_save_on_store_1", self.store_id == 1, |_| {});
+
+            let now = Instant::now();
+            self.perf_context.start_observe();
+            self.engines
+                .raft
+                .consume_and_shrink(
+                    &mut self.batch.raft_wb,
+                    false,
+                    RAFT_WB_SHRINK_SIZE,
+                    RAFT_WB_DEFAULT_SIZE,
+                )
+                .unwrap_or_else(|e| {
+                    panic!(
+                        "store {}: {} failed to write to raft engine: {:?}",
+                        self.store_id, self.tag, e
+                    );
+                });
+            self.perf_context.report_metrics();
+            STORE_WRITE_RAFTDB_DURATION_HISTOGRAM
+                .observe(duration_to_sec(now.saturating_elapsed()) as f64);
+        }
+
+        self.batch.after_write_to_db(&self.metrics);
+
+        self.last_unpersisted_id += 1;
```

Comment:
Why plus one here? `self.last_persisted_id < self.global_persisted_id.load(Ordering::Acquire)` needs to wait for global_persisted_id was increased by 2?

Reply:
This is `last_unpersisted_id` not `last_persisted_id`. `last_unpersisted_id` is just a monotonic increasing counter and is used for recording the last unpersisted id.